polkit/validate/validate.go

```diff
@@ -0,0 +1,290 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package validate
+
+import (
+	"encoding/xml"
+	"fmt"
+	"io"
+	"strings"
+)
+
+type Element struct {
+	CharData        string     `xml:",chardata"`
+	UnknownAttrs    []xml.Attr `xml:",any,attr"`
+	UnknownChildren []xml.Name `xml:",any"`
+}
+
+type policyConfig struct {
+	XMLName xml.Name `xml:"policyconfig"`
+	Element
+
+	Vendor    []Element `xml:"vendor"`
+	VendorURL []Element `xml:"vendor_url"`
+	IconName  []Element `xml:"icon_name"`
+
+	Actions []action `xml:"action"`
+}
+
+type action struct {
+	Element
+
+	ID string `xml:"id,attr"`
+
+	Vendor    []Element `xml:"vendor"`
+	VendorURL []Element `xml:"vendor_url"`
+	IconName  []Element `xml:"icon_name"`
+
+	Description []message  `xml:"description"`
+	Message     []message  `xml:"message"`
+	Defaults    []defaults `xml:"defaults"`
+	Annotate    []annotate `xml:"annotate"`
+}
+
+type message struct {
+	Element
+
+	GettextDomain string `xml:"gettext-domain,attr"`
+	Language      string `xml:"lang,attr"` // to match xml:lang
+}
+
+type defaults struct {
+	Element
+
+	AllowAny      []Element `xml:"allow_any"`
+	AllowInactive []Element `xml:"allow_inactive"`
+	AllowActive   []Element `xml:"allow_active"`
+}
+
+type annotate struct {
+	Element
+
+	Key string `xml:"key,attr"`
+}
+
+func ValidatePolicy(r io.Reader) (actionIDs []string, err error) {
+	decoder := xml.NewDecoder(r)
+	var config policyConfig
+	if err := decoder.Decode(&config); err != nil {
+		return nil, err
+	}
+	// check for additional data after the root element
+	switch err := decoder.Decode(new(interface{})); err {
+	case nil:
+		return nil, fmt.Errorf("invalid XML: additional data after root element")
+	case io.EOF:
+		// nothing
+	default:
+		return nil, err
+	}
```

Comment:
Nitpick: in case the valid XML is followed by some garbage, the returned error might be misleading. I'd rather rewrite this:
```suggestion
	if err := decoder.Decode(new(interface{})); err != io.EOF {
		return nil, fmt.Errorf("invalid XML: additional data after root element")
	}
```

Reply:
This seems like a nice simplification. I tried adding a test case for extra character data after the root element, but that seems to return `io.EOF` error as no extra data. I've added a test with an additional incomplete element after the root, which is covered by your suggested change.